src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -641,3 +649,295 @@ ExtractPartitionValue(Query *query, Var *partitionColumn)
+	task->dependedTaskList = NIL;
+	task->upsertQuery = upsertQuery;
+
+	return task;
+}
+
+
+/*
+ * RouterQueryJob creates a Job for the specified query to execute the
+ * provided single shard select task.*/
+static Job *
+RouterQueryJob(Query *query, Task *task)
+{
+	Job *job = NULL;
+	List *taskList = NIL;
+	TaskType taskType = task->taskType;
+
+	/*
+	 * We send modify task to the first replica, otherwise we choose the target shard
+	 * according to task assignment policy.
+	 */
+	if (taskType == MODIFY_TASK)
+	{
+		taskList = FirstReplicaAssignTaskList(list_make1(task));
+	}
+	else
+	{
+		taskList = AssignAnchorShardTaskList(list_make1(task));
+	}
+
+	job = CitusMakeNode(Job);
+	job->dependedJobList = NIL;
+	job->jobId = INVALID_JOB_ID;
+	job->subqueryPushdown = false;
+	job->jobQuery = query;
+	job->taskList = taskList;
+
+	return job;
+}
+
+
+/*
+ * MultiRouterPlannableQuery returns true if given query can be router plannable.
+ * The query is router plannable if it is a select query issued on a hash
+ * partitioned distributed table, and it has a exact match comparison on the
+ * partition column. This feature is enabled if task executor is set to real-time or
+ * router.
+ */
+bool
+MultiRouterPlannableQuery(Query *query)
+{
+	uint32 rangeTableId = 1;
+	List *rangeTableList = NIL;
+	RangeTblEntry *rangeTableEntry = NULL;
+	Oid distributedTableId = InvalidOid;
+	Var *partitionColumn = NULL;
+	char partitionMethod = '\0';
+	Node *quals = NULL;
+	CmdType commandType = query->commandType;
+	FromExpr *joinTree = query->jointree;
+	List *varClauseList = NIL;
+	ListCell *varClauseCell = NULL;
+	bool partitionColumnMatchExpression = false;
+	int partitionColumnReferenceCount = 0;
+	int shardCount = 0;
+	bool inTransaction = false;
+
+	Assert(commandType == CMD_SELECT);
+
+	/*
+	 * Reject subqueries which are in SELECT or WHERE clause.
+	 * Queries which are recursive, with CommonTableExpr, with locking (hasForUpdate),
+	 * or with window functions are also rejected here.
+	 * Queries which have subqueries, or tablesamples in FROM clauses are rejected later
+	 * during RangeTblEntry checks.
+	 */
+	if (query->hasSubLinks == true || query->cteList != NIL || query->hasForUpdate ||
+		query->hasRecursive)
+	{
+		return false;
+	}
+
+#if (PG_VERSION_NUM >= 90500)
+	if (query->groupingSets)
+	{
+		return false;
+	}
+#endif
+
+	/* only hash partitioned tables are supported */
+	distributedTableId = ExtractFirstDistributedTableId(query);
+	partitionColumn = PartitionColumn(distributedTableId, rangeTableId);
+	partitionMethod = PartitionMethod(distributedTableId);
+
+	if (partitionMethod != DISTRIBUTE_BY_HASH)
+	{
+		return false;
+	}
+
+	/* extract range table entries */
+	ExtractRangeTableEntryWalker((Node *) query, &rangeTableList);
+
+	/* query can have only one range table of type RTE_RELATION */
+	if (list_length(rangeTableList) != 1)
+	{
+		return false;
+	}
+
+	rangeTableEntry = (RangeTblEntry *) linitial(rangeTableList);
```

Comment:
Minor: I am curious why do we have this check? On what query we return false here?